### ISSUE TITLE
Operator sometimes does not apply all updates

### ISSUE THREAD

**Issue by pstewy** [dragonflydb/dragonfly-operator#93]
When working with the operator, I found there are times where it thinks the pods it is managing match the statefulset, but that is not the case. An example of this is applying an update to the installed CRD instance that updates the replicas and the memory limit. If I increase the replicas, then the operator will spin up the new pods, but never update the already existing pods. 

While doing some troubleshooting, I found that adding a few second sleep [here](https://github.com/dragonflydb/dragonfly-operator/blob/main/internal/controller/dragonfly_controller.go#L236) fixes the problem (which I figure is not an acceptable fix). It appears the operator retrieves the statefulset before the cluster has fully acknowledged the change.

**Comment by Pothulapati** [comment]
Ah, So a CRD update that includes both replicas change along with memory limit? Maybe the rollout mechanism is not getting triggered as we expect to. Will investigate! 👍🏼 

sleep is probably not an acceptable fix, and we should think of something better here! Also, Change ack is usually pretty fast right? wonder why its not working as expected 🤔

**Comment by pstewy** [comment]
Yeah sleep definitely is not ideal and does not (to no one's surprise 😄 ) fix all the cases. My understanding is the ctrl client by default uses a cache, so its possible the [request](https://github.com/dragonflydb/dragonfly-operator/blob/main/internal/controller/dragonfly_controller.go#L239) to get the statefulset returns the cached object. I tried comparing the spec of the created one (which I pulled before applying the update to the cluster) vs the existing one (taken from [this thread](https://github.com/kubernetes-sigs/kubebuilder/issues/592)) and my initial testing shows it working as expected. 
Another method I've read is taking a hash of the spec and adding it to the pods as an annotation. Then when checking whether we need to do a rollout or not we compare the hash on the pods vs. the hash of the changes that were just applied. This covers the case where the operator is restarted for whatever reason between applying the changes to the cluster and triggering the rollout.
